src/main/java/io/github/opencubicchunks/cubicchunks/levelgen/aquifer/CubicAquifer.java

```diff
@@ -91,7 +91,16 @@ private int getIndex(int x, int y, int z) {
         int localX = x - this.minGridX;
         int localY = y - this.minGridY;
         int localZ = z - this.minGridZ;
-        return (localY * this.gridSizeZ + localZ) * this.gridSizeX + localX;
+
+        int index = (localY * this.gridSizeZ + localZ) * this.gridSizeX + localX;
+
+        //FIXME: Because of the way the cache is implemented, this can cause an ArrayIndexOutOfBoundsException.
+        //The coordinates are unpacked from longs so when outside the proper range, they wrap around and this returns nonsense values.
```

Comment:
Does this happen in cubic or vanilla worlds? If it only happens in vanilla, this check should ensure that we are in a cubic world and raise an exception instead.

Reply:
The reason BlockPos wraps around within vanilla bounds is because of [../mixin/core/common/MixinBlockPos](https://github.com/OpenCubicChunks/CubicChunks2/blob/a83bb3acf7221ecbaf9b7d492d0fde94967a4158/src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinBlockPos.java) which results in less bits being used in a long to represent those coordinates whatever the world type. Therefore, this wraparound will most likely occur within any world using CubicAquifer.